pkg/iac/terraform/state/backend/tfcloud_reader.go

```diff
@@ -26,38 +28,79 @@ type TFCloudBody struct {
 	Data TFCloudData `json:"data"`
 }
 
-func NewTFCloudReader(client pkghttp.HTTPClient, workspaceId string, opts *Options) (*HTTPBackend, error) {
-	req, err := http.NewRequest(http.MethodGet, fmt.Sprintf("%s/workspaces/%s/current-state-version", TFCloudAPI, workspaceId), nil)
-
-	if err != nil {
-		return nil, err
-	}
-
-	req.Header.Add("Content-Type", "application/vnd.api+json")
-	req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", opts.TFCloudToken))
-
-	res, err := client.Do(req)
+type TFCloudBackend struct {
+	request *http.Request
+	client  pkghttp.HTTPClient
+	reader  io.ReadCloser
+}
 
+func NewTFCloudReader(client pkghttp.HTTPClient, workspaceId string, opts *Options) (*TFCloudBackend, error) {
+	req, err := http.NewRequest(http.MethodGet, fmt.Sprintf("%s/workspaces/%s/current-state-version", TFCloudAPI, workspaceId), nil)
 	if err != nil {
 		return nil, err
 	}
 
-	if res.StatusCode < 200 || res.StatusCode >= 400 {
-		return nil, errors.Errorf("error requesting terraform cloud backend state: status code: %d", res.StatusCode)
+	token := opts.TFCloudToken
+	if token == "" {
+		tfConfigFile, err := getTerraformConfigFile()
+		if err != nil {
+			return nil, err
+		}
+		file, err := os.Open(tfConfigFile)
+		if err != nil {
+			return nil, err
+		}
+		reader := NewTFCloudConfigReader(file)
+		token, err = reader.GetToken()
+		if err != nil {
+			return nil, err
+		}
 	}
 
-	bodyBytes, _ := ioutil.ReadAll(res.Body)
+	req.Header.Add("Content-Type", "application/vnd.api+json")
+	req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", token))
 
-	body := TFCloudBody{}
-	err = json.Unmarshal(bodyBytes, &body)
+	return &TFCloudBackend{req, client, nil}, nil
+}
 
-	if err != nil {
-		return nil, err
+func (t *TFCloudBackend) Read(p []byte) (n int, err error) {
+	if t.reader == nil {
+		res, err := t.client.Do(t.request)
+		if err != nil {
+			return 0, err
+		}
+
+		if res.StatusCode < 200 || res.StatusCode >= 400 {
+			return 0, errors.Errorf("error requesting terraform cloud backend state: status code: %d", res.StatusCode)
+		}
+
+		body := TFCloudBody{}
+		bodyBytes, _ := ioutil.ReadAll(res.Body)
+		err = json.Unmarshal(bodyBytes, &body)
+		if err != nil {
+			return 0, err
+		}
+
+		rawURL := body.Data.Attributes.HostedStateDownloadUrl
+		logrus.WithFields(logrus.Fields{"hosted-state-download-url": rawURL}).Trace("Terraform Cloud backend response")
+
+		h, err := NewHTTPReader(t.client, rawURL, &Options{})
+		if err != nil {
+			return 0, err
+		}
+		n, err = h.Read(p)
+		if err != nil {
+			return 0, err
+		}
+		t.reader = h.reader
+		return n, nil
```

Comment:
This method could be simplified

```go
func (t *TFCloudBackend) Read(p []byte) (n int, err error) {
	if t.reader == nil {
       // ...
		h, err := NewHTTPReader(t.client, rawURL, &Options{})
		if err != nil {
			return 0, err
		}
        // h implement ReadCloser
        // how could you assign h.reader field since it is unexported ?
		t.reader = h 
        // No need to return here
	}
	return t.reader.Read(p)
}
```

Reply:
Good catch! I didn't think to just assign `h` ...

> how could you assign h.reader field since it is unexported ?

It is in the same package (e.g. `backend`).